### PR TITLE
Add RoI align operator.

### PR DESCRIPTION
**Pull request by jerrywgz**

### CODE REVIEW

File: paddle/fluid/operators/roi_align_op.cc [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        ctx.device_context());
+  }
+};
+
+class ROIAlignOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X",
+             "(Tensor), "
+             "the input of ROIAlignOp. "
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.cc]: 和其他保持一致，首字母都大写吧，the input -> The

File: paddle/fluid/operators/roi_align_op.cc [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+             "The format of input tensor is NCHW. Where N is batch size, "
+             "C is the number of input channels, "
+             "H is the height of the feature, and "
+             "W is the width of the feature.");
+    AddInput("ROIs",
+             "(LoDTensor), "
+             "ROIs (Regions of Interest) to pool over. "
+             "should be a 2-D LoDTensor of shape (num_rois, 4)"
+             "given as [[x1, y1, x2, y2], …]. "
+             "Where batch_id is the id of the data, "
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.cc]: There is no batch_id in ROIs, remove `Where batch_id is the id of the data...`

File: paddle/fluid/operators/roi_align_op.cu [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+  return std::min((N + kNumCUDAThreads - 1) / kNumCUDAThreads,
+                  kNumMaxinumNumBlocks);
+}
+
+#define CUDA_1D_KERNEL_LOOP(i, n)                              \
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < (n); \
+       i += blockDim.x * gridDim.x)
+
+template <class T>
+__device__ T bilinear_interpolate(const T* input_data, const int height,
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.cu]: bilinear_interpolate -> BilinearInterpolate

File: paddle/fluid/operators/roi_align_op.cu [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+  T v3 = input_data[y_high * width + x_low];
+  T v4 = input_data[y_high * width + x_high];
+  T w1 = hy * hx, w2 = hy * lx, w3 = ly * hx, w4 = ly * lx;
+
+  T val = (w1 * v1 + w2 * v2 + w3 * v3 + w4 * v4);
+  return val;
+}
+
+template <class T>
+__device__ void bilinear_interpolate_gradient(const int height, const int width,
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.cu]: bilinear_interpolate_gradient -> BilinearInterpolateGradient

File: paddle/fluid/operators/roi_align_op.cu [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+    int rois_num_with_lod = rois_lod[rois_batch_size];
+    PADDLE_ENFORCE_EQ(rois_num, rois_num_with_lod,
+                      "The rois_num from input and lod must be the same.");
+    for (int n = 0; n < rois_batch_size; ++n) {
+      for (size_t i = rois_lod[n]; i < rois_lod[n + 1]; ++i) {
+        roi_batch_id_data[i] = n;
+      }
+    }
+    Tensor roi_batch_id_list_gpu;
+    framework::TensorCopy(roi_batch_id_list, ctx.GetPlace(),
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.cu]: use TensorCopySync

File: paddle/fluid/operators/roi_align_op.cu [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+          roi_batch_id_list.mutable_data<int>(platform::CPUPlace());
+      auto rois_lod = rois->lod().back();
+      int rois_batch_size = rois_lod.size() - 1;
+      for (int n = 0; n < rois_batch_size; ++n) {
+        for (size_t i = rois_lod[n]; i < rois_lod[n + 1]; ++i) {
+          roi_batch_id_data[i] = n;
+        }
+      }
+      Tensor roi_batch_id_list_gpu;
+      framework::TensorCopy(roi_batch_id_list, ctx.GetPlace(),
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.cu]: same as above

File: paddle/fluid/operators/roi_align_op.h [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+using LoDTensor = framework::LoDTensor;
+
+static constexpr int kROISize = 4;
+
+template <class T>
+void pre_calc_for_bilinear_interpolate(
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.h]: The name style does not follow Google C++ code style.

pre_calc_for_bilinear_interpolate -> PreCalcForBilinearInterpolate

File: paddle/fluid/operators/roi_align_op.cu [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+    auto pooled_width = ctx.Attr<int>("pooled_width");
+    auto spatial_scale = ctx.Attr<float>("spatial_scale");
+    auto sampling_ratio = ctx.Attr<int>("sampling_ratio");
+
+    int rois_num = rois->dims()[0];
+    int channels = in->dims()[1];
+    int height = in->dims()[2];
+    int width = in->dims()[3];
+
+    if (in_grad) {
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.cu]:
```c++
if (in_grad) {
  return;
}
xxx
```

Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.cu]:
```c++
if (!in_grad) {
  return;
}
xxx
```

File: paddle/fluid/operators/roi_align_op.h [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+            output_val /= count;
+            output_data[pool_index] = output_val;
+          }
+        }
+        batch_data += in_stride[1];
+        output_data += out_stride[1];
+      }
+      rois_data += roi_stride[0];
+    }
+    return;
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.h]: remove this line.

File: paddle/fluid/operators/roi_align_op.h [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+    auto* out_grad =
+        ctx.Input<framework::Tensor>(framework::GradVarName("Out"));
+    auto* in_grad = ctx.Output<framework::Tensor>(framework::GradVarName("X"));
+
+    auto pooled_height = ctx.Attr<int>("pooled_height");
+    auto pooled_width = ctx.Attr<int>("pooled_width");
+    auto spatial_scale = ctx.Attr<float>("spatial_scale");
+    auto sampling_ratio = ctx.Attr<int>("sampling_ratio");
+    auto in_dims = in->dims();
+    if (in_grad) {
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.h]: same as above.

File: paddle/fluid/operators/roi_align_op.h [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+                                                batch_grad_data);
+                }
+              }
+            }
+          }
+        }
+        rois_data += roi_stride[0];
+      }
+    }
+    return;
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.h]: remove this line.

File: python/paddle/fluid/layers/nn.py [hunk 12]
```diff
@@ -5177,6 +5178,53 @@ def roi_pool(input, rois, pooled_height=1, pooled_width=1, spatial_scale=1.0):
     return pool_out
 
 
+@templatedoc()
+def roi_align(input,
+              rois,
+              pooled_height=1,
+              pooled_width=1,
+              spatial_scale=1.0,
+              sampling_ratio=-1):
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: add name:
```python
# ...
sampling_ratio=-1,
name=None):

```

File: paddle/fluid/operators/roi_align_op.h [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+            }
+            pre_calc_index += 1;
+            continue;
+          }
+          if (y <= 0) {
+            y = 0;
+          }
+          if (x <= 0) {
+            x = 0;
+          }
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.h]:
```c++
y = y <= 0 ? 0 : y;
x = x <= 0 ? 1 : x;
```

File: paddle/fluid/operators/roi_align_op.h [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+    w1 = w2 = w3 = w4 = 0;
+    x_low = x_high = y_low = y_high = -1;
+    return;
+  }
+  if (y <= 0) {
+    y = 0;
+  }
+  if (x <= 0) {
+    x = 0;
+  }
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.h]:
```c++
y = y <= 0 ? 0 : y;
x = x <= 0 ? 0 : x;
```

File: paddle/fluid/operators/roi_align_op.h [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+  T diff2 = out_grad_this_bin * w2 / count;
+  T diff3 = out_grad_this_bin * w3 / count;
+  T diff4 = out_grad_this_bin * w4 / count;
+  if (x_low >= 0 && x_high >= 0 && y_low >= 0 && y_high >= 0) {
+    *(batch_grad_data + y_low * width + x_low) += diff1;
+    *(batch_grad_data + y_low * width + x_high) += diff2;
+    *(batch_grad_data + y_high * width + x_low) += diff3;
+    *(batch_grad_data + y_high * width + x_high) += diff4;
+  }
+  return;
```
Comment (qingqing01) [inline paddle/fluid/operators/roi_align_op.h]: remove this line.

File: paddle/fluid/operators/roi_align_op.cc (hunk 1 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.cc]: Done

File: paddle/fluid/operators/roi_align_op.cc (hunk 2 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.cc]: Done

File: paddle/fluid/operators/roi_align_op.cu (hunk 3 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.cu]: Done

File: paddle/fluid/operators/roi_align_op.cu (hunk 4 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.cu]: Done

File: paddle/fluid/operators/roi_align_op.cu (hunk 5 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.cu]: Done

File: paddle/fluid/operators/roi_align_op.cu (hunk 8 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.cu]: Done

File: paddle/fluid/operators/roi_align_op.cu (hunk 6 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.cu]: Done

File: paddle/fluid/operators/roi_align_op.h (hunk 7 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.h]: Done

File: paddle/fluid/operators/roi_align_op.h (hunk 13 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.h]: Done

File: paddle/fluid/operators/roi_align_op.h (hunk 14 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.h]: Done

File: paddle/fluid/operators/roi_align_op.h (hunk 15 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.h]: Done

File: paddle/fluid/operators/roi_align_op.h (hunk 9 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.h]: Done

File: paddle/fluid/operators/roi_align_op.h (hunk 10 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.h]: Done

File: paddle/fluid/operators/roi_align_op.h (hunk 11 above)
Comment (jerrywgz) [reply paddle/fluid/operators/roi_align_op.h]: Done

File: python/paddle/fluid/layers/nn.py (hunk 12 above)
Comment (jerrywgz) [reply python/paddle/fluid/layers/nn.py]: Done

**Review by qingqing01 — APPROVED** [review]

**Review by xingyuanbu — APPROVED** [review]

**Review by xingyuanbu — APPROVED** [review]
approve

File: python/paddle/fluid/layers/nn.py [hunk 16]
```diff
@@ -5177,6 +5178,54 @@ def roi_pool(input, rois, pooled_height=1, pooled_width=1, spatial_scale=1.0):
...
+        pooled_width (integer): ${pooled_width_comment} Default: 1
+        spatial_scale (float): ${spatial_scale_comment} Default: 1.0
+        sampling_ratio(intger): ${sampling_ratio_comment} Default: -1
+
+    Returns:
+        Variable: ${out_comment}.
+    Examples:
+        .. code-block:: python
+
+            align_out = fluid.layers.roi_align(input=x,
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:5205]: Can you give a more detailed example, such as examples of x with value [xxx], rois with value [xxx]...many sequence_xxx op has very detailed examples

**Review by panyx0718 — APPROVED** [review]
LG API, can have better example

**Pull request merged**